### PR TITLE
Evaluate GPU halving on recent history

### PR DESCRIPTION
**Pull request by enarjord**
## Summary
- change opt-in single-coin Trailing Martingale successive halving from historical prefixes to recent-history suffixes
- give each partial scoring window the normal indicator warmup immediately before its trade boundary
- keep the full-history rung unchanged and version the checkpoint contract so prefix-era checkpoints fail closed

## Validation
- `PYO3_PYTHON=venv/bin/python cargo test --manifest-path passivbot-rust/Cargo.toml --no-default-features gpu::tests --lib` (14 passed)
- `pytest tests/optimization/test_gpu_mps.py` on Apple MPS
- `pytest tests/optimization/test_gpu_service.py tests/optimization/test_gpu_backend.py`
- CPU import isolation check confirms ordinary CPU startup does not import torch or the MPS kernel
- bounded opt-in GPU optimize smoke completed one 1024-candidate generation across recent 25%, recent 50%, and full-history rungs, then shut down cleanly via SIGINT

## Contract
- exact Rust validation remains authoritative
- the feature remains disabled by default
- EMA Anchor, multicoin, suite, CPU optimization, backtest, and bot paths are unchanged

### CODE REVIEW

File: src/optimization/gpu/mps_kernel.py [hunk 1]
```diff
@@ -3040,7 +3117,9 @@ def dispatch_once():
             )
         )
         if self.recovery_distribution_enabled:
-            output["strategy_eq_recovery_samples"] = recovery_samples
+            output["strategy_eq_recovery_samples"] = recovery_samples[
+                :, :effective_recovery_sample_count
+            ]
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/mps_kernel.py:3204]: Fixed in bd2233e09. The shader now uses the full `n_recovery_samples` capacity from `sizes[9]` for every physical row offset and a separate bounded logical count from `sizes[12]` for write bounds and the returned slice. The recent-window parity test now runs two candidates with recovery sampling enabled and compares their samples against an independently sliced run.

File: passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal [hunk 2]
```diff
@@ -1974,7 +1979,8 @@ inline void passivbot_single_coin_impl(
     );
 #endif
 
-    for (int k = 1; k < T - 1; ++k) {
+    const int loop_start = recent_history_window ? max(seed_k + 1, 1) : 1;
+    for (int k = loop_start; k < T - 1; ++k) {
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal:2008]: Fixed in bd2233e09. Trailing Martingale now packs an hour-boundary bit into the existing hourly-valid flag and, for bounded runs, rebuilds each hourly range and carried latest value from the bounded warmup seed. This preserves the 11-column flag ABI and adds no per-candle traffic. The sliced-run parity test now starts at a non-hour boundary with nonzero 1h-volatility weights on two candidates.

**Comment by enarjord** [comment]
@codex review

File: src/optimization/gpu/service.py [hunk 3]
```diff
@@ -2098,29 +2105,46 @@ def _parameter_matrix(self, candidates: list[dict]) -> np.ndarray:
...
+        history_start = max(
+            int(self.run.first_valid_idx),
+            trade_start
+            - max(
+                1,
+                int(
+                    getattr(
+                        self, "history_warmup_bars", self.run.warmup_bars
+                    )
+                ),
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/service.py]: Fixed in ca5708e8e. Recent-window routing now reserves `max(1, warmup_bars) + 1` candles before the scoring boundary, matching the ordinary `indices > max(1, warmup_bars)` readiness contract. It also will not choose a trade start earlier than that boundary relative to the first valid candle. The sliced MPS parity test now uses `warmup_bars=20` with trade index 21, so an off-by-one generation would fail.

File: src/optimization/backends/gpu_backend.py [hunk 4]
```diff
@@ -3979,12 +3979,16 @@ def evaluate_proxy(candidates, *, history_fraction=1.0):
...
         def evaluate_proxy(candidates, *, history_fraction=1.0):
-            end_step = (
-                proxy.end_step_for_history_fraction(history_fraction)
-                if float(history_fraction) < 1.0
-                else None
-            )
-            return proxy.evaluate(candidates, end_step=end_step)
+            if float(history_fraction) < 1.0:
+                history_start, trade_start = (
+                    proxy.recent_window_for_history_fraction(history_fraction)
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/backends/gpu_backend.py:3985]: Fixed in ca5708e8e. Option normalization now rewrites the accepted terminal fraction to exactly `1.0` immediately after tolerance validation, before policy storage, checkpoint identity, or evaluation routing. A regression test configures `0.9999999999995` and asserts the resolved ladder ends at exact `1.0`.

**Comment by enarjord** [comment]
Addressed the `ca5708e8e` terminal-profile finding in `96e6983cb`. Bounded profiling now subtracts the history-window start before clamping, estimating avoidable work, and computing terminal p50/p90 fractions. Added a regression with an absolute step 75 inside a 50–100 window; it asserts 25/48 progress and 23 remaining candidate-bars. The full GPU service module passes.

File: src/optimization/gpu/service.py [hunk 5]
```diff
@@ -2098,29 +2105,48 @@ def _parameter_matrix(self, candidates: list[dict]) -> np.ndarray:
...
+        )
+        full_trade_start = min(
+            candle_count - 3,
             max(
-                3,
-                int(self.run.trade_start_idx) + 2,
-                int(self.run.first_valid_idx) + 2,
+                2,
+                int(self.run.trade_start_idx),
+                int(self.run.first_valid_idx) + warmup_readiness_bars + 1,
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/service.py]: Fixed in 4fb266dbd. `full_trade_start` again derives from the ordinary run baseline (`trade_start_idx`, with only the existing first-valid safety floor); the strict extra readiness candle is applied only when choosing the bounded `history_start`. Added a late-listing regression with `first_valid_idx=900`, warmup 100, and `trade_start_idx=1000`, asserting the full mapping preserves trade start 1000. The complete GPU service module passes.

File: passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal [hunk 6]
```diff
@@ -1974,21 +1980,69 @@ inline void passivbot_single_coin_impl(
...
+            && (!recent_history_window || k >= bounded_trade_start);
         const int di = flags[fo + 2];
 #if !PASSIVBOT_TM_VOLATILITY_DISABLED
-        const bool hour_valid = flags[fo + 3] != 0;
+        const int hour_flags = flags[fo + 3];
+        bool hour_valid = (hour_flags & 1) != 0;
+        if (recent_history_window) {
+            const bool hour_boundary = (hour_flags & 2) != 0;
+            if (hour_boundary) {
+                if (k > bounded_hour_start_k + 1 && bounded_hour_has_price
```
[the remark this answers is not in the document]
Comment (enarjord) [reply passivbot-rust/src/gpu/mps_trailing_martingale_directional.metal]: Fixed in 03af5601a. The bounded TM path now follows the same timestamp-based hourly readiness and overlapping-window rules as the CPU materializer, including adjacent boundaries for intervals that do not divide one hour. Added sliced-vs-bounded MPS parity for a suffix beginning at 00:45 with 45-minute candles; the focused regression, full MPS module, rebuilt extension, and all 14 Rust GPU source-contract tests pass.
